### ISSUE TITLE
High resource usage

### ISSUE THREAD

**Issue by InfamousVague** [Satellite-im/WarpGUI#36]
Investigate ways we can bring down resource usage. Right now we're using quite a bit of resources and we want WarpGUI to run on any system so we'll need to bring this down. This issue involves looking through the render loop and seeing what could be causing these issues 

<img width="554" alt="image" src="https://user-images.githubusercontent.com/1770198/192360734-bce6e785-5c07-4cb8-9657-9d5fc976e24c.png">

**Comment by dariusc93** [comment]
As discussed, this is due to Friends and Profile component being looped over after logging in or creating an account, which is using the high cpu usage. Temporary solution would be to add a small delay in those components, but other solution would be to only render those components when they are active/shown.

**Comment by InfamousVague** [comment]
Or we could render the component conditionally so that it's only looping when you're using it, in addition to the delay for now

**Comment by InfamousVague** [comment]
https://github.com/Satellite-im/WarpGUI/pull/84

**Comment by dariusc93** [comment]
To leave a note here regarding my investigation into the performance issue:

Previously during my first check, it was pertaining to `Friends` and `Profile` components being the issue for the performance, however alot has changed since the first time ive checked it out and even adding a temporary delay. After further investigating last night and today, it was concluded that the following components were being rerendered:

- Write
- Messages
- Friends (when it is opened)
- Profile 
- Popup 
- ActivityIndicator

There may be others that may cause it not listed here

Digging into it further and reviewing the documentation and code from dioxus, it looks like most of the hit happens because of direct calls to functions to manipulate the vdom rather than altering the state itself, which is causing the vdom to be re-rendered, with it not being the calls itself causing the performance issue (eg calls directly to warp). Take Profile as an example, The main reasoning for it rerendering is due to it using 

```rust
    let friends = use_state(&cx, Vec::new);
    friends.set(mp.read().list_friends().unwrap_or_default());
```
Here, we are setting the state after using `use_state` and causing it to be rerendered. Removing the second line and changing the first line to use either `Vec::<DID>::new` or `|| mp.read().list_friends().unwrap_or_default()` would stop it from being rerendered. Now the question would then be how can be update that state? Well the solution here as I mentioned before would be to use `cx.spawn` and pass the reference there so we can alter it without it automatically impacting the dom. In this case, we can do the following

```rust
    let friends_inner = friends.clone();
    let mp_inner = mp.clone();
    cx.spawn(async move {
        loop {
            let list = mp_inner.read().list_friends().unwrap_or_default();
            if *friends_inner != list {
                friends_inner.set(list);
            }
            tokio::time::sleep(std::time::Duration::from_secs(1)).await;
        }
    });

```
To quickly explain whats going on here, we are cloning the reference of multipass and the friends state and passing it into the spawned task. Here we are looping to check the friends list and if there is any changes to the list, we would set the state to use the new list. Of course the duration can be lower (even low as 100ms would be fine), but a second would do for now. The same can apply for `ActivityIndicator` where if you move the if condition into a task, it would drop it from being rerendered unless the state is changed then it would only render for that change. Eg

```rust
    let status = use_state(&cx, || IdentityStatus::Offline);

    let account = cx.props.account.clone();
    let status_state = status.clone();
    let remote_did = cx.props.remote_did.clone();

    cx.spawn(async move {
        loop {
            if let Ok(current_status) = account.read().identity_status(&remote_did) {
                if *status_state != current_status {
                    status_state.set(current_status);
                }
            }
            tokio::time::sleep(std::time::Duration::from_secs(1)).await;
        }
    });
``` 

Components like `Write` issue lies in the `script` where it is setting the state to use the script causing this to be rerendered without any updates. Something like this might be more suited to be in a const or just a regular string variable
```rust
    let script = r#"
        (function addAutoResize() {
            document.querySelectorAll('.resizeable-textarea').forEach(function (element) {
                let send_button = document.getElementById('send');
                send_button.addEventListener('click', function(event) {
                    element.value = '';
                });

                element.addEventListener('keyup', function(event) {
                    if (event.keyCode === 13 && !event.shiftKey) {
                        event.target.value = '';
                        event.target.style.height = 'auto';
                    }
                });

                element.style.boxSizing = 'border-box';
                var offset = element.offsetHeight - element.clientHeight;
                element.addEventListener('input', function (event) {
                    event.target.style.height = 'auto';
                    event.target.style.height = event.target.scrollHeight + offset + 'px';
                });
                element.removeAttribute('data-autoresize');
            });
        })()"#;
```

Messages component might need rethinking as messages arent being stored in a state but being fetched directly. In this case, a `cx.spawn` could be used but would only allow messages to be rendered only once after loading the component since it is not fetching it from the future and restarting it. 

Popups (et al) may need more investigating to see what would be triggering their rerender

Notes (may get updated):
- `use_coroutine` could also be used in this case instead as it would likely give a little more control over when you can start and stop it
- Warp has an events/hooks system but due to changes made months ago as well as how things progressed, it needs refactoring which would come after things are more stable and a better design is in place (see https://github.com/Satellite-im/Warp/issues/34). Even then, due to how dioxus is designed, an external event system wouldnt directly help with the performance since it may still require spawning a separate future task or due to how things are designed in the app.
- Make use of states (or use_ref) where you expect values to change but avoid changing the state directly after setting but only when there is need for change (eg updating input, etc)
- Set values upon initial load (eg friend list) directly in `use_state` then update accordingly from there.
- For components that need to update specific states using a function call should use `spawn` or `use_coroutine` where needed.
- Review dioxus docs and source code to understand how it handles the state of the components. 
- There might be a chance that spawning task would leak as I have not seen anything where the scope would cancel it when the component is unmounted.

Avg usage (around 4% to 15%) after logging in with *some* changes done (the changes I've done wont get pushed up right now since it might need more thought in how everything is layout and not everything is fixed, though may get pushed to a separate branch but the info above should be enough to get somebody started in understanding on how to possibly proceed, but wouldnt mind providing guidance and advice):
![image](https://user-images.githubusercontent.com/3918024/193918945-21c573c3-0f8d-4434-b863-f99c571503ee.png)

**Comment by dariusc93** [comment]
Note: After testing, I caught onto something. When the page rerenders, it doesnt cancel the previous task that is running so that task would run while the new task is spawned up, causing higher cpu usage over time. Based on the code I was reading and documentation, it should have been cancelling the task when the component is unmounted, but it doesnt look to be the case.

Workaround may be to track the taskid for the future task spawned after rerendering to check for the previous taskid for that component and cancel it before spawning a new one, or not to spawn one in if it exist for that scope. Another temporary solution would be not to loop in the task but instead process everything there and when the component rerenders it would spawn in a task but the other would have been marked completed after being polled and seeing it is ready instead of pending. 

Still requires abit of research in how dioxus handles task and if this been fixed or resolved upstream or not (since we are using a older version of the library due to some issues when attempting to use the repo instead of the crates version - maybe been resolved?)

**Comment by dariusc93** [comment]
Just another side note: It looks like the cx.spawn acts a true task (which I guess make sense) so I did replace what ive done to use `use_future` since it does perform a check on that scope for any existing task and remove it before creating a new one. Messages component is refactored a bit (though conversation switching is broken) and last that is left is popup component but I will leave that to @InfamousVague to resolve
